include/cetl/span.h

```diff
@@ -1 +1,772 @@
+    /// @see std::span::data
+    constexpr pointer data() const noexcept
+    {
+        return data_;
+    }
+
+    ///@}
+    // +----------------------------------------------------------------------+
+    ///@{ @name Observers
+    // +----------------------------------------------------------------------+
+
+    ///
+    /// If the span has a zero size or not.
+    /// @return true if the span size is 0 where "size" is the same as span::extent
+    /// for this specialization.
+    [[nodiscard]] constexpr bool empty() const noexcept
+    {
+        return (Extent == 0);
+    }
+
+    ///
+    /// The size of the span.
+    /// @return The number of elements in the span which is equal to `span::extent`.
+    /// @see std::span::size
+    constexpr size_type size() const noexcept
+    {
+        return Extent;
+    }
+
+    ///
+    /// The size of the span in bytes.
+    /// @return `sizeof(element_type) * size_`.
+    /// @see std::span::size_bytes
+    constexpr size_type size_bytes() const noexcept
+    {
+        return sizeof(element_type) * Extent;
+    }
+    ///@}
+    // +----------------------------------------------------------------------+
+    ///@{ @name Subviews
+    // +----------------------------------------------------------------------+
+    /// Create a new span from the start of the current span for `Count` elements.
+    /// @tparam Count   Number of elements for the sub-span.
+    /// @return A new span with the static extent `Count`.
+    /// @see std::span::first
+    template <size_type Count>
+    constexpr span<element_type, Count> first() const
+    {
+        static_assert(Count <= Extent, "subviews beyond the size of the span's view are not allowed.");
+        return span<element_type, Count>(data_, Count);
+    }
+
+    /// Create a new span from the start of the current span for count elements.
+    /// @param count  The number of elements for the sub-span. Where `count` is greater than the
+    ///               current span's size the behavior is undefined.
+    /// @return A new span with a dynamic extent extent and a size of `count`.
+    /// @see std::span::first
+    constexpr span<element_type, cetl::dynamic_extent> first(size_type count) const
+    {
+        CETL_DEBUG_ASSERT(count <= Extent, "Dynamic subviews beyond the size of the span's view are undefined.");
+        return span<element_type, cetl::dynamic_extent>(data_, count);
+    }
+
+    /// Create a new span `Count` elements from the last item of the current span to its end.
+    /// @tparam Count   Number of elements for the sub-span.
+    /// @return A new span with the static extent `Count`.
+    /// @see std::span::last
+    template <size_type Count>
+    constexpr span<element_type, Count> last() const
+    {
+        static_assert(Count <= extent, "subviews beyond the size of the span's view are not allowed.");
+        return span<element_type, Count>{(data_ + extent) - Count, Count};
```

Comment:
would prefer index math over pointer math.

Reply:
both are prohibited by AUTOSAR C++ though

Reply:
index math is more clear and in a single "domain" instead of having an implied `*sizeof(X)`

Reply:
Okay.